File: src/main/java/kr/choyunjin/chlorine/i18n/I18n.java
package kr.choyunjin.chlorine.i18n;

import java.io.InputStream;
import java.io.IOException;
import org.tomlj.TomlParseResult;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;

public class I18n {
    private MiniMessage mm;
    private TomlParseResult data;
    
    public I18n(TomlParseResult data) {
        this.mm = MiniMessage.miniMessage();
        this.data = data;
    }
    
    public Component tl(String key, TagResolver... params) {
        return this.mm.deserialize(this.data.getString(key), params);
    }

    // minimessage 문법이 없는 경우 사용
    public String tlraw(String key, TagResolver... params) {
        return this.mm.serialize(this.tl(key, params));
    }
    
    public TagResolver param(String name, Component value) {
        return Placeholder.component(name, value);
    }

    public TagResolver param(String name, String value) {
        return Placeholder.unparsed(name, value);
    }
}

File: src/main/kotlin/kr/choyunjin/chlorine/i18n/I18n.kt
package kr.choyunjin.chlorine.i18n

import org.json.JSONObject
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage

private fun getValueRecursively(jsonObj: JSONObject, key: List<String>): String? {
    val obj = jsonObj.get(key.get(0))
    if (key.size > 1) {
        if (obj is JSONObject) {
            return getValueRecursively(obj as JSONObject, key.subList(1, key.size))
        } else return null
    } else if (obj is String) {
        return obj as String
    } else return null
}

class I18n(val translations: JSONObject) {
    private fun getTranslatedString(key: String): String {
        val tlValue = getValueRecursively(this.translations, key.split("."));
        return tlValue ?: key;
    }

    fun tl(key: String): Component {
        return MiniMessage.miniMessage().deserialize(this.getTranslatedString(key));
    }

    fun tl(key: String, initContext: TranslationContext.() -> Unit): Component {
        val ctx = TranslationContext()
        ctx.initContext()
        return MiniMessage.miniMessage().deserialize(this.getTranslatedString(key), *ctx.params.toTypedArray());
    }

    // minimessage 문법이 없는 경우 사용
    fun tlraw(key: String,): String {
        return MiniMessage.miniMessage().serialize(this.tl(key));
    }

    fun tlraw(key: String, initContext: TranslationContext.() -> Unit): String {
        return MiniMessage.miniMessage().serialize(this.tl(key, initContext));
    }
}